File: 携程8.23笔试/t2.py
import sys, math

def select_primes(upper, flags, primes: list):
    for i in range(2, upper):
        if not flags[i]:
            primes.append(i)
        for pj in primes:
            if pj > upper // i:
                break
            flags[pj * i] = True
            if i % pj == 0:
                break

def resolve(x, primes):
    res = []
    for pi in primes:
        if x % pi == 0:
            res.append(pi)
            while x % pi == 0:
                x //= pi
    # 记得不要漏掉x本身是质数的情况
    if x > 1:
        res.append(x)
    return res

def read_int():
    with open('input.txt', 'r', encoding='utf-8') as file:
        data = file.read().strip().split()
        for i in data:
            yield int(i)

def main():
    it = iter(read_int())
    n = next(it, 0)
    print(n)
    
    freqs = {}
    _max = 0
    # 没必要获取全部的输入，只要拿到>1的那部分就行了
    for _ in range(n):
        num = next(it)
        if num <= 1:
            continue
        freqs[num] = freqs.get(num, 0) + 1
        _max = max(_max, num)

    if len(freqs.keys()) == 0:
        print("0 0")
        sys.exit(0)

    # 筛质数
    upper = math.isqrt(_max) + 1
    prime_flag = [False] * (upper + 1)
    primes: list[int] = []
    select_primes(upper, prime_flag, primes)

    # 分解质因数，求最大长度
    cover = {} # 每个质数能覆盖几个元素
    base = {} # 每个质数覆盖的元素组当中，各个元素的gcd是多少
    for n, t in freqs.items():
        for pi in resolve(n, primes):
            cover[pi] = cover.get(pi, 0) + t
            base[pi] = n if pi not in base else math.gcd(base[pi], n)

    ans = max(cover.values(), default=1)
    if ans == 1:
        print(f"1 {len(freqs)}")
        sys.exit(0)

    # 统计答案
    seen = set()
    cnt = 0
    candidates = [pi for pi in cover if cover[pi] == ans]
    for pi in candidates:
        if pi in seen:
            continue
        for qi in resolve(base[pi], primes):
            seen.add(qi)
        cnt += 1
    
    print(f"{ans} {cnt}")



if __name__ == "__main__":
    main()

File: 携程8.23笔试/t2.cpp
#include <cmath>
#include <cstdio>
#include <iostream>
#include <algorithm>
#include <array>
#include <vector>
#include <unordered_map>
#include <unordered_set>
constexpr int N = 2e6 + 5, M = 2e8 + 5;
int P = static_cast<int>(std::sqrt(M));

std::unordered_map<int, int> freqs, cover, base;
std::vector<int> primes, factors;
int n;

inline void select_primes(int upper) {
    std::vector<char> flag (upper + 1, 0);
    for (int i = 2; i <= upper; i++) {
        if (!flag[i])
            primes.emplace_back(i);
        for (int j = 0; j < primes.size() && primes[j] <= upper / i; j++) {
            flag[primes[j] * i] = true;
            if (i % primes[j] == 0)
                break;
        }
    }
}

void resolve(int x) {
    factors.clear();
    for (auto pi: primes) {
        if (x % pi == 0) {
            factors.push_back(pi);
            while (x % pi == 0) {
                x /= pi;
            }
        }
    }
    if (x != 1) {
        factors.push_back(x);
    }
}

int gcd(int x, int y) {
    if (!y) {
        return x;
    }
    return gcd(y, x%y);
}

int main() {
    std::freopen("input.txt", "r", stdin);
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    // 没必要获取全部输入，只要>1的
    int x, _max = 0;
    std::cin >> n;
    for (int i = 1; i <= n; i++) {
        std::cin >> x;
        if (x <= 1)
            continue;
        freqs[x]++;
        _max = std::max(_max, x);
    }
    if(freqs.empty()) {
        std::cout << "0 0";
        return 0;
    }
    int upper = static_cast<int>(std::sqrt(_max)) + 1;
    // 筛质数
    select_primes(upper);
    // 统计长度
    int L = 0;
    for (auto &pi: freqs) {
        int key = pi.first;
        resolve(key);
        for (auto ai: factors) {
            cover[ai] += pi.second;
            if (base.find(ai) == base.end()) {
                base[ai] = pi.first;
            } else {
                base[ai] = gcd(base[ai], pi.first);
            }
        }
    }
    for (auto &pi: cover) {
        L = std::max(L, pi.second);
    }
    if (L == 1) {
        std::cout << L << " " << freqs.size() << std::endl;
        return 0;
    }
    std::unordered_set<int> seen;
    int ans = 0;
    // 注意这里是cover而不是primes
    for (auto &pi: cover) {
        if (seen.contains(pi.first)) 
            continue;
        if (cover[pi.first] < L) {
            continue;
        }
        resolve(base[pi.first]);
        for (auto &vi: factors) {
            seen.insert(vi);
        }
        ans++;
    }
    std::cout << L << " " << ans << std::endl;
    return 0;
}